packs/newrelic/sensors/newrelic_application_sensor.py

```diff
@@ -0,0 +1,311 @@
+        # make sure all servers are ok.
+        all_servers_ok = True
+        for server in servers:
+            all_servers_ok &= server['health_status'] in ['green']
+            if not all_servers_ok:
+                break
+
+        if all_servers_ok:
+            self._sensor_service.dispatch(WEB_APP_NORMAL_TRIGGER_REF, payload)
+        else:
+            for server in servers:
+                self._log.info('server %s has state %s. Rescheduling normal check.',
+                               server['name'], server['health_status'])
+            eventlet.spawn_after(self._normal_report_delay, self._dispatch_server_normal,
+                                 payload, attempt_no + 1)
+
+    # alert test methods
+    def _is_alert_opened(self, long_description):
+        return long_description and long_description.startswith('Alert opened')
+
+    def _is_alert_closed(self, long_description):
+        return long_description and long_description.startswith('Alert ended')
+
+    def _is_alert_canceled(self, long_description):
+        return long_description and long_description.startswith('Alert canceled')
+
+    def _is_alert_acknowledged(self, long_description):
+        return long_description and long_description.startswith('Alert acknowledged')
+
+    def _is_escalated_downtime(self, long_description):
+        return long_description and long_description.startswith('Alert escalated to downtime')
+
+    def _is_downtime_recovered(self, long_description):
+        return long_description and long_description.startswith('Alert downtime recovered')
+
+    # newrelic API methods
+    def _get_application(self, app_name):
+        params = None
+        if app_name:
+            params = {'filter[name]': app_name}
+        url = urllib_parse.urljoin(self._api_url, 'applications.json')
+        resp = requests.get(url, headers=self._headers, params=params).json()
+        if 'applications' in resp:
+            # pick 1st application
+            return resp['applications'][0] if resp['applications'] else None
+        return None
+
+    def _get_servers(self, server_names):
+        servers = {}
+        # No batch query by name support so making API calls in a tight loop. Might be
+        # ok to get all severs and filter manually but that gets complex for a large number
+        # of server since the API pages data.
+        for server_name in server_names:
+            params = {'filter[name]': server_name}
+            url = urllib_parse.urljoin(self._api_url, 'servers.json')
+            resp = requests.get(url, headers=self._headers, params=params).json()
+            servers['server_name'] = resp['servers'][0] if resp['servers'] else None
+        return servers
+
+    def cleanup(self):
+        # self._log.info('cleanup called.')
+        # # If Flask is using the default Werkzeug server, then call shutdown on it.
+        # func = request.environ.get('werkzeug.server.shutdown')
```

Comment:
I tried to put in cleanup code but it ends up throwing exceptions. Shall remove since this is now in the commit history and I can bring it back if needed.